src/theory/sep/theory_sep.cpp

```diff
@@ -1744,105 +1747,128 @@ bool TheorySep::areDisequal( Node a, Node b ){
+  {
+    return;
+  }
+  // allocate the heap assert info for e1
+  HeapAssertInfo* e1 = getOrMakeEqcInfo(t1, true);
+  std::vector<Node> toAdd[2];
+  for (size_t i = 0; i < 2; i++)
+  {
+    bool pol = i == 0;
+    NodeList& e2list = pol ? e2->d_posPto : e2->d_negPto;
+    for (const Node& p : e2list)
+    {
+      if (checkPto(e1, p, pol))
+      {
+        // add unless checkPto determined it was redundant
+        toAdd[i].push_back(p);
       }
     }
-    e1->d_has_neg_pto.set( e1->d_has_neg_pto.get() || e2->d_has_neg_pto.get() );
-    //validate
-    validatePto( e1, t1 );
+  }
+  // now that checks are complete, add them all now
+  for (size_t i = 0; i < 2; i++)
+  {
+    NodeList& e1list = i == 0 ? e1->d_posPto : e1->d_negPto;
+    for (const Node& p : toAdd[i])
+    {
+      e1list.push_back(p);
+    }
   }
 }
 
-void TheorySep::validatePto( HeapAssertInfo * ei, Node ei_n ) {
-  if( !ei->d_pto.get().isNull() && ei->d_has_neg_pto.get() ){
-    for( NodeList::const_iterator i = d_spatial_assertions.begin(); i != d_spatial_assertions.end(); ++i ) {
-      Node fact = (*i);
-      if (fact.getKind() == kind::NOT)
+bool TheorySep::checkPto(HeapAssertInfo* e, Node p, bool polarity)
+{
+  Assert(e != nullptr);
+  Assert(p.getKind() == SEP_LABEL && p[0].getKind() == SEP_PTO);
+  NodeManager* nm = NodeManager::currentNM();
+  Node plbl = p[1];
+  Node pval = p[0][1];
+  bool ret = true;
+  // check for inferences involving p with all pto constraints already
+  // contained in e.
+  for (size_t i = 0; i < 2; i++)
+  {
+    bool pol = i == 0;
+    NodeList& elist = pol ? e->d_posPto : e->d_negPto;
+    for (const Node& q : elist)
+    {
+      Assert(q.getKind() == SEP_LABEL && q[0].getKind() == SEP_PTO);
+      Node qlbl = q[1];
+      Node qval = q[0][1];
+      // We use instantiated labels where labels are set to singletons. We
+      // assume these always share a root label.
+      if (qlbl.getKind() != SET_SINGLETON && plbl.getKind() != SET_SINGLETON
+          && !sharesRootLabel(plbl, qlbl))
+      {
+        Trace("sep-pto") << "Constraints " << p << " and " << q
+                         << " do not share a root label" << std::endl;
+        // if do not share a parent, skip
+        continue;
+      }
+      if (polarity && pol)
       {
-        TNode atom = fact[0];
-        Assert(atom.getKind() == kind::SEP_LABEL);
-        TNode satom = atom[0];
-        if (satom.getKind() == SEP_PTO)
+        // two positive pto
+        if (!areEqual(pval, qval))
         {
-          if( areEqual( atom[1], ei_n ) ){
-            addPto( ei, ei_n, atom, false );
+          std::vector<Node> exp;
+          if (plbl != qlbl)
+          {
+            Assert(areEqual(plbl, qlbl));
+            exp.push_back(plbl.eqNode(qlbl));
+          }
+          exp.push_back(p);
+          exp.push_back(q);
+          // enforces injectiveness of pto
+          //  (pto x y) ^ (pto y w) ^ x = y => y = w
```

Comment:
My understanding locations and values can have different types. So `y` here appears as a value in `(pto x y)` and as a location in `(pto y w)`. Is this a typo?

Reply:
Also, the premise `x = y` here equates between locations, but the code equates between labels `plbl = qlbl` which are sets.

Reply:
Yes this is a typo, I'll fix.